homeassistant/components/geniushub/__init__.py

```diff
@@ -13,25 +13,38 @@
 
 DOMAIN = 'geniushub'
 
+_V1_API_SCHEMA = vol.Schema({
+    vol.Required(CONF_TOKEN): cv.string,
+})
+
+_V3_API_SCHEMA = vol.Schema({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+})
+
 CONFIG_SCHEMA = vol.Schema({
-    DOMAIN: vol.Schema({
-        vol.Optional(CONF_USERNAME): cv.string,
-        vol.Optional(CONF_PASSWORD): cv.string,
-        vol.Required(CONF_HOST): cv.string,
-    }),
+    DOMAIN: vol.Any(
+        _V3_API_SCHEMA,
+        _V1_API_SCHEMA,
+    )
 }, extra=vol.ALLOW_EXTRA)
 
 
 async def async_setup(hass, hass_config):
     """Create a Genius Hub system."""
     from geniushubclient import GeniusHubClient  # noqa; pylint: disable=no-name-in-module
 
-    host = hass_config[DOMAIN].get(CONF_HOST)
-    username = hass_config[DOMAIN].get(CONF_USERNAME)
-    password = hass_config[DOMAIN].get(CONF_PASSWORD)
-
     geniushub_data = hass.data[DOMAIN] = {}
 
+    if hass_config[DOMAIN].get(CONF_HOST):
+        host = hass_config[DOMAIN].get(CONF_HOST)
+        username = hass_config[DOMAIN].get(CONF_USERNAME)
+        password = hass_config[DOMAIN].get(CONF_PASSWORD)
+    else:
+        host = hass_config[DOMAIN].get(CONF_TOKEN)
```

Comment:
Don't call this host, when it's not a host.

Eg collect the arguments in a tuple and pass the tuple to the class.

Reply:
Which do you prefer:
```python
    host = token = None
    if CONF_HOST in hass_config[DOMAIN]:
        host = hass_config[DOMAIN].get(CONF_HOST)
        username = hass_config[DOMAIN].get(CONF_USERNAME)
        password = hass_config[DOMAIN].get(CONF_PASSWORD)
    else:
        token = hass_config[DOMAIN].get(CONF_TOKEN)
        username = password = None

    try:
        client = geniushub_data['client'] = GeniusHubClient(
            host if host else token, username, password,
            session=async_get_clientsession(hass)
        )
```

or
```python
    if CONF_HOST in hass_config[DOMAIN]:
        hub_identifier = hass_config[DOMAIN].get(CONF_HOST)
        username = hass_config[DOMAIN].get(CONF_USERNAME)
        password = hass_config[DOMAIN].get(CONF_PASSWORD)
    else:
        hub_identifier = hass_config[DOMAIN].get(CONF_TOKEN)
        username = password = None

    try:
        client = geniushub_data['client'] = GeniusHubClient(
            hub_identifier, username, password,
            session=async_get_clientsession(hass)
        )
```

Reply:
Collect the arguments in a tuple.
```py
if blabla:
    args = arg1, arg2, etc
else:
    args = (arg4, )
client = class(*args, session=session)
```

Reply:
Oh, I see - _but_ it would a tuple of 1 (hub_id), the rest are kwargs/dict.  I don't see what would be gained:
```python
def __init__(self, hub_id, username=None, password=None, session=None, debug=False)
```
So how about:
```python
    kwargs = dict(hass_config[DOMAIN])
    if CONF_HOST in kwargs:
        args = (kwargs.pop(CONF_HOST), )
    else:
        args = (kwargs.pop(CONF_TOKEN), )

    try:
        client = geniushub_data['client'] = GeniusHubClient(
            *args, **kwargs, session=async_get_clientsession(hass)
        )
```

Reply:
We gain that we don't confuse ourselves by naming a variable something which it isn't.

Reply:
👍 In addition, I think it's kind of elegant - you get a two-for-one from voluptuous.